app/lib/slug_router.rb

```diff
@@ -0,0 +1,66 @@
+# rubocop:disable Style/StringLiterals
+class SlugRouter
+  attr_reader :handle, :slug
+
+  class << self
+    def call(request)
+      new(**request.parameters.symbolize_keys)
+    end
+
+    alias [] call
+
+    def logger
+      Rails.logger
+    end
+  end
+
+  # TODO: We should eventually refactor these routes to use :handle in place
+  # of :username — as it's currently used for pages, organizations, users, etc.
+  def initialize(handle: nil, username: nil, slug: nil, **_options)
+    @handle = handle || username # can transition gradually
+    @slug = slug
+  end
+
+  def map
+    Rails.cache.fetch(File.join(*["routing", handle, slug].compact), expires_in: 12.hours) do
+      map_without_cache
+    end
+  end
+
+  def map_without_cache
+    slug.blank? ? map_for_handle_only : map_for_handle_with_slug
+  end
+
+  def map_for_handle_only
+    if Page.exists?(slug: handle, is_top_level_path: true)
+      logger.info " ]]]]]> SlugRouter: Top-level Page!"
+      'pages#show'
+    elsif Podcast.exists?(slug: handle)
+      logger.info " ]]]]]> SlugRouter: Podcasts#show!"
+      'podcasts#show'
+    elsif Organization.exists?(username: handle)
+      logger.info " ]]]]]> SlugRouter: Organizations#show"
+      'organizations#show'
+    elsif User.exists?(username: handle)
+      logger.info " ]]]]]> SlugRouter: Users#show"
+      'users#show'
+    end
+  end
+
+  def map_for_handle_with_slug
+    if Article.exists?(path: "/#{handle.downcase}/#{slug}")
```

Comment:
As I see, we'll first use `SlugRouter` only for `PodcastEpisode`s but we'll still search for Articles - so more queries.
Maybe that's ok because there is caching.

Reply:
I believe that it should be similar to the queries we're doing now, even before caching, and significantly fewer queries after caching. For reference, the code [this would eventually be replacing](https://github.com/forem/forem/blob/2c138d86ef97e0a532ef2151d297c779d74f4d86/app/controllers/stories_controller.rb#L38-L44) currently on `main` does:

```ruby
if (@article = Article.includes(:user).find_by(path: path)&.decorate)
  handle_article_show
elsif (@article = Article.find_by(slug: params[:slug])&.decorate)
  handle_possible_redirect
elsif (@podcast = Podcast.available.find_by(slug: params[:username]))
  @episode = @podcast.podcast_episodes.available.find_by!(slug: params[:slug])
  handle_podcast_show
else
  not_found
end
```

Differences:
* The existing code does a full Article query (including a join with `user`) before any/every podcast episode lookup
* The existing code does an additional full article query to [possibly redirect](https://github.com/forem/forem/blob/2c138d86ef97e0a532ef2151d297c779d74f4d86/app/controllers/stories_controller.rb#L84-L98) before any/every podcast episode lookup (the redirect behavior isn't currently covered in this PR, but I don't think it ought to come _before_ every PodcastEpisode in any case).
* The new code will sometimes do an extraneous `Article.exists?` query (in the sense that `articles#show` still needs to do a full `find`) but the "mapping" for that handle/slug path should then be cached.